concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/window/ImmutablePercentileSampleWindow.java

```diff
@@ -15,51 +15,52 @@
  */
 package com.netflix.concurrency.limits.limit.window;
 
-import java.util.ArrayList;
-import java.util.Comparator;
-import java.util.List;
+import java.util.Arrays;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLongArray;
 
 class ImmutablePercentileSampleWindow implements SampleWindow {
     private final long minRtt;
     private final int maxInFlight;
     private final boolean didDrop;
-    private final List<Long> observedRtts;
+    private final AtomicLongArray observedRtts;
+    private final int sampleCount;
     private final double percentile;
 
-    ImmutablePercentileSampleWindow(double percentile) {
+    ImmutablePercentileSampleWindow(double percentile, int windowSize) {
         this.minRtt = Long.MAX_VALUE;
         this.maxInFlight = 0;
         this.didDrop = false;
-        this.observedRtts = new ArrayList<>();
+        this.observedRtts = new AtomicLongArray(windowSize);
+        this.sampleCount = 0;
         this.percentile = percentile;
     }
 
     ImmutablePercentileSampleWindow(
             long minRtt,
             int maxInFlight,
             boolean didDrop,
-            List<Long> observedRtts,
+            AtomicLongArray observedRtts,
+            int sampleCount,
             double percentile
     ) {
         this.minRtt = minRtt;
         this.maxInFlight = maxInFlight;
         this.didDrop = didDrop;
         this.observedRtts = observedRtts;
+        this.sampleCount = sampleCount;
         this.percentile = percentile;
     }
 
     @Override
     public ImmutablePercentileSampleWindow addSample(long rtt, int inflight, boolean didDrop) {
-        // TODO: very naive
```

Comment:
Need to check if `sampleCount < windowSize` and ignore the sample.  You'd want to still take didDrop and inflight into account.

Reply:
you mean when `sampleCount > windowSize`?

Reply:
> You'd want to still take didDrop and inflight into account.

To be honest, I have no clue how number of inflight requests and didDrop might affect windowedLimit flow(other than just passing this to underlying limit) since every latency, including erroneous are accounted after https://github.com/Netflix/concurrency-limits/pull/126
can you provide any thoughts?

Reply:
@elandau kindly ping

Reply:
We should definitely put an upper bound on the number of latency measurements to avoid having to allocate excessive memory.  Dropping samples is fine since percentiles have the nice property of being accurate when sampled (I read this somewhere but couldn't find links to this)

Reply:
both are fixed + removed heuristic from https://github.com/Netflix/concurrency-limits/pull/119#discussion_r284339788